File: problems/CSES/Graph_Algorithms/counting_rooms/counting_rooms.py
import sys
from collections import deque
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n, m = map(int, input().split())
    grid = [list(input().strip()) for _ in range(n)]

    rooms = 0

    directions = [
        (1, 0),
        (-1, 0),
        (0, 1),
        (0, -1)
    ]

    for r in range(n):
        for c in range(m):
            if grid[r][c] != '.':
                continue

            rooms += 1

            q = deque()
            q.append((r, c))
            grid[r][c] = '#'

            while q:
                row, col = q.popleft()

                for dr, dc in directions:
                    nr = row + dr
                    nc = col + dc

                    if (
                            0 <= nr < n
                            and 0 <= nc < m
                            and grid[nr][nc] == '.'
                    ):
                        grid[nr][nc] = '#'
                        q.append((nr, nc))

    print(rooms)

main()

File: problems/CSES/Graph_Algorithms/counting_rooms/counting_rooms.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int n, m;
    cin >> n >> m;

    vector<string> grid(n);
    for (int i = 0; i < n; i++) cin >> grid[i];

    int rooms = 0;

    vector<pair<int, int>> directions = {
        {1, 0},
        {-1, 0},
        {0, 1},
        {0, -1},
    };

    for (int r = 0; r < n; r++) {
        for (int c = 0; c < m; c++) {
            if (grid[r][c] != '.')
                continue;

            rooms++;

            deque<pair<int, int>> q;
            q.push_back({r, c});
            grid[r][c] = '#';

            while (!q.empty()) {
                auto [row, col] = q.front();
                q.pop_front();

                for (auto [dr, dc] : directions) {
                    int nr = row + dr;
                    int nc = col + dc;

                    if (nr >= 0 && nr < n && nc >= 0 && nc < m && grid[nr][nc] == '.') {
                        grid[nr][nc] = '#';
                        q.push_back({nr, nc});
                    }
                }
            }
        }
    }

    cout << rooms << '\n';

    return 0;
}
